sql/job_stat_history_log_retention.sql

```diff
@@ -4,18 +4,128 @@
+  INTO drop_after
+  FROM _timescaledb_config.bgw_job
+  WHERE id = 3;
+
+  IF NOT FOUND THEN
+    RAISE EXCEPTION 'system job 3 (job execution history retention) not found';
+  END IF;
+
+  RAISE DEBUG 'search_point %, drop_after %',
+    search_point, drop_after;
+
+  SELECT COALESCE(min(id), 0), COALESCE(max(id), 0)
+  INTO id_lower, id_upper
+  FROM _timescaledb_internal.bgw_job_stat_history;
+
+  IF id_lower = 0 AND id_upper = 0 THEN
+    RETURN NULL;
+  END IF;
+
+  -- We want the first entry in the table where execution_finish is >= search_point
+  WHILE id_lower < id_upper LOOP
+    id_middle := id_lower + (id_upper - id_lower) / 2;
+
+    SELECT execution_finish
+    INTO target_tz
+    FROM _timescaledb_internal.bgw_job_stat_history
+    WHERE id = id_middle;
+
+    -- If the id_middle is not found, shift to the next id
+    IF NOT FOUND THEN
+      SELECT execution_finish, id
+      INTO target_tz, id_middle_next
+      FROM _timescaledb_internal.bgw_job_stat_history
+      WHERE id >= id_middle AND id <= id_upper
+      ORDER BY id LIMIT 1;
+
+      IF NOT FOUND THEN
+        RETURN NULL;
+      END IF;
+
+      RAISE DEBUG 'id_middle % not found, shift to the next id %', id_middle, id_middle_next;
+      id_middle := id_middle_next;
+    END IF;
+
+    RAISE DEBUG 'target_tz %, id_lower %, id_upper %, id_middle %, le %, gt %',
+      target_tz, id_lower, id_upper, id_middle, (target_tz <= search_point),
+      (target_tz > search_point);
+
+    IF search_point >= target_tz THEN
+      id_upper := id_middle;
+    ELSE
+      id_lower := id_middle + 1;
+    END IF;
+  END LOOP;
+
+  RETURN id_lower;
+END;
+$$
+LANGUAGE plpgsql SET search_path TO pg_catalog, pg_temp, public;
+
 CREATE OR REPLACE FUNCTION _timescaledb_functions.policy_job_stat_history_retention(job_id integer, config JSONB) RETURNS integer
 LANGUAGE PLPGSQL AS
 $BODY$
 DECLARE
     drop_after INTERVAL;
     numrows INTEGER;
+    search_point TIMESTAMPTZ;
+    id_found BIGINT;
 BEGIN
     drop_after := config->>'drop_after';
+  SELECT now() - drop_after::interval
+  INTO search_point
+  FROM _timescaledb_config.bgw_job
+  WHERE id = 3;
 
-    DELETE
-    FROM _timescaledb_internal.bgw_job_stat_history
-    WHERE execution_finish < (now() - drop_after);
+  IF NOT FOUND THEN
+    RAISE EXCEPTION 'system job 3 (job execution history retention) not found';
+    RETURN 0;
+  END IF;
+
+  id_found := job_history_bsearch(search_point);
+
+  IF id_found IS NULL THEN
+    RAISE WARNING 'no job history for cleaning up';
+    RETURN 0;
+  END IF;
+
+  CREATE TEMP TABLE __tmp_bgw_job_stat_history ON COMMIT DROP AS
+  SELECT * FROM _timescaledb_internal.bgw_job_stat_history
+  WHERE id >= id_found AND execution_finish >= search_point
+  ORDER BY id;
+
+  ALTER TABLE _timescaledb_internal.bgw_job_stat_history
+    DROP CONSTRAINT IF EXISTS bgw_job_stat_history_pkey;
+
+  DROP INDEX IF EXISTS _timescaledb_internal.bgw_job_stat_history_job_id_idx;
```

Comment:
```suggestion
```
I think we shouldn't do DDL inside this job cause is easy to forget to change it if for some reason in the future we change something in this table